gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/DataFlowTopology.java

```diff
@@ -0,0 +1,124 @@
+package gobblin.data.management.copy.replication;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+import com.typesafe.config.Config;
+
+import gobblin.data.management.copy.CopyableDataset;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/**
+ * Class to represent the data flow topology from copy destination to copy sources
+ * @author mitu
+ *
+ */
+@AllArgsConstructor
+public class DataFlowTopology {
+
+  public static final String ROUTES = "routes";
+  
+  @Getter
+  private final List<CopyRoute> routes;
+
+  private DataFlowTopology(Builder builder) {
+    Preconditions.checkArgument(builder.source != null, "Can not build topology without source");
+    Preconditions.checkArgument(builder.replicas != null && builder.replicas.size() > 0, "Can not build topology without replicas");
+    Preconditions.checkArgument(builder.topologyConfig != null, "Can not build topology without topology config");
+    Preconditions.checkArgument(builder.topologyConfig.hasPath(ROUTES), "Can not build topology without " + ROUTES);
+
+    // key of the map is replication name, value is {@link ReplicationReplica}
+    Map<String, ReplicationReplica> replicasMap = new HashMap<String, ReplicationReplica>();
+    for (ReplicationReplica replica : builder.replicas) {
+      String name = replica.getReplicationName();
+      // replica name can not be {@link ReplicationUtils#REPLICATION_SOURCE}
+      Preconditions.checkArgument(!name.equals(ReplicationUtils.REPLICATION_SOURCE),
+          String.format("replica name %s can not be reserved word %s ", name, ReplicationUtils.REPLICATION_SOURCE));
+      replicasMap.put(name, replica);
+    }
+
+    this.routes = new ArrayList<CopyRoute>();
+
+    Config routesConfig = builder.topologyConfig.getConfig(ROUTES);
+    // routes should be available for each replica
+    for (ReplicationReplica replica : builder.replicas) {
+      Preconditions.checkArgument(routesConfig.hasPath(replica.getReplicationName()),
+          "can not find route for replica " + replica.getReplicationName());
+
+      CopyRoute route = new CopyRoute();
+      route.copyDestination = replica;
+      route.copyFrom = new ArrayList<ReplicationData<CopyableDataset>>();
+
+      List<String> copyFroms = routesConfig.getStringList(replica.getReplicationName());
+
+      for (String copyFromName : copyFroms) {
+        Preconditions.checkArgument(!copyFromName.equals(route.copyDestination.getReplicationName()),
```

Comment:
Maybe change this to a warning and only throw error if there are no valid `copyFroms`?